### PR TITLE
switch check_type() to use cli package

### PR DESCRIPTION
**Pull request by EmilHvitfeldt**
to close https://github.com/tidymodels/recipes/issues/1217.

This PR switches `check_type()` to use {cli} in its errors. Additionally and more importantly, it lists the offending variables and their types 🎉 

``` r
library(recipes)
library(modeldata)

recipe(~., data = iris) |>
  step_date(all_predictors()) |>
  prep()
#> Error in `step_date()`:
#> Caused by error in `prep()`:
#> ✖ All columns selected for the step should be date, or datetime.
#> • 4 double variables found: `Sepal.Length`, `Sepal.Width`, …
#> • 1 factor variable found: `Species`

recipe(~., data = ames) |>
  step_date(all_predictors()) |>
  prep()
#> Error in `step_date()`:
#> Caused by error in `prep()`:
#> ✖ All columns selected for the step should be date, or datetime.
#> • 40 factor variables found: `MS_SubClass`, `MS_Zoning`, `Street`, …
#> • 12 double variables found: `Lot_Frontage`, `Mas_Vnr_Area`, …
#> • 22 integer variables found: `Lot_Area`, `Year_Built`, …
```

### CODE REVIEW

File: R/misc.R [hunk 1]
```diff
@@ -436,13 +436,36 @@ check_type <- function(dat, quant = TRUE, types = NULL, call = caller_env()) {
...
-        "All columns selected for the step",
-        " should be ",
-        label,
-        "."
-      ),
+    info <- get_types(dat)[!all_good, ]
+    classes <- map_chr(info$type, function(x) x[1])
+    counts <- vctrs::vec_split(info$variable, classes)
+    counts$count <- lengths(counts$val)
+    counts$text_len <- cli::console_width() - 19 - (counts$count > 1) -
```
Comment (EmilHvitfeldt) [inline R/misc.R:443]: `-19` - my count of the length of "•   variable found: "
`(counts$count > 1)` - whether variable has an `s` or not
`nchar(counts$key)` - length of the variable type

File: R/misc.R [hunk 2]
```diff
@@ -436,13 +436,36 @@ check_type <- function(dat, quant = TRUE, types = NULL, call = caller_env()) {
...
+      nchar(counts$key)
+
+    # cli::ansi_collapse() doesn't work for length(x) == 1
+    # https://github.com/r-lib/cli/issues/590
+    variable_collapse <- function(x, width) {
+      x <- paste0("{.var ", x, "}")
+      if (length(x) == 1) {
+        res <- cli::ansi_strtrim(x, width = width)
```
Comment (EmilHvitfeldt) [inline R/misc.R:451]: in the unlikely event that the single variable name is too large, i used `cli::ansi_strtrim` to truncate the variable by itself

File: R/misc.R [hunk 3]
```diff
@@ -436,13 +436,36 @@ check_type <- function(dat, quant = TRUE, types = NULL, call = caller_env()) {
...
+    }
+
+    problems <- glue::glue_data(
+      counts,
+      "{count} {key} variable{ifelse(count == 1, '', 's')} \\
+      found: {purrr::map2_chr(val, text_len, variable_collapse)}"
+    )
+    names(problems) <- rep("*", length(problems))
+
+    message <- "All columns selected for the step should be {label}."
```
Comment (simonpcouch) [inline R/misc.R:465]:
```suggestion
    message <- "All columns selected for the step should be {.or {types}}."
```

Some of the relevant lines are outside of where I'm allowed to comment/suggest, but I noticed that there's an additional comma added when there are only two options in `glue::glue_collapse(types, sep = ", ", last = ", or ")`. e.g., from the snaps for `step_cut()`:

> `x All columns selected for the step should be double, or integer.`

cli has a `.or` inline markup option that would come in handy here.

Besides this, these are all that would need to change, I think:

Screenshot 2023-10-11 at 10 26 24 AM

File: tests/testthat/_snaps/factor2string.md [hunk 4]
```diff
@@ -5,7 +5,8 @@
     Condition
       Error in `step_factor2string()`:
       Caused by error in `prep()`:
-      ! All columns selected for the step should be factor, or ordered.
+      x All columns selected for the step should be factor, or ordered.
+      * 2 string variables found: `w`, and `x`
```
Comment (simonpcouch) [inline tests/testthat/_snaps/factor2string.md:9]: No need for a comma here.

File: R/misc.R [hunk 5]
```diff
@@ -436,13 +436,36 @@ check_type <- function(dat, quant = TRUE, types = NULL, call = caller_env()) {
...
+      } else {
+        res <- cli::ansi_collapse(x, width = width, style = "head")
+      }
+      res
+    }
+
+    problems <- glue::glue_data(
+      counts,
+      "{count} {key} variable{ifelse(count == 1, '', 's')} \\
+      found: {purrr::map2_chr(val, text_len, variable_collapse)}"
```
Comment (simonpcouch) [inline R/misc.R:461]: I believe we can write out `variable_collapse()` altogether, instead writing:

```suggestion
      found: {paste0('{.var ', val, '}')}"
```

This would resolve the current issue where a comma is placed in two-item lists.

This does indeed mean that the single, long variable name is not truncated, instead line-wrapped to the console. I'm okay with that edge case for the sake of reduced complexity.

**Review by simonpcouch — CHANGES_REQUESTED** [review]
This is a big improvement! Feel free to re-request review when you feel this is in a good spot. :)

File: tests/testthat/_snaps/factor2string.md (hunk 4 above)
Comment (EmilHvitfeldt) [reply tests/testthat/_snaps/factor2string.md:9]: fixed by using `.or` inline markup

File: R/misc.R (hunk 5 above)
Comment (EmilHvitfeldt) [reply R/misc.R:461]: The suggested change results in:

Screenshot 2023-10-11 at 9 47 39 AM

I wrote `variable_collapse()` because I wanted to use `ansi_collapse()` so we only print out one line of variables instead of all of them

File: R/misc.R (hunk 3 above)
Comment (EmilHvitfeldt) [reply R/misc.R]: closed in [`d962daa` (#1235)](https://github.com/tidymodels/recipes/pull/1235/commits/d962daaa49a3c001e6b423412ff80d40970d1d86)

File: tests/testthat/_snaps/factor2string.md (hunk 4 above)
Comment (EmilHvitfeldt) [reply tests/testthat/_snaps/factor2string.md:9]: closed in [`d962daa` (#1235)](https://github.com/tidymodels/recipes/pull/1235/commits/d962daaa49a3c001e6b423412ff80d40970d1d86)

File: R/misc.R (hunk 5 above)
Comment (EmilHvitfeldt) [reply R/misc.R:466]: I removed the comma for two item list now! in [`9106ffd` (#1235)](https://github.com/tidymodels/recipes/pull/1235/commits/9106ffd0be561ee04acb253669346afcc0886a3f)

``` r
library(recipes)
library(modeldata)

recipe(~., data = iris) |>
  step_date(Sepal.Length, Sepal.Width, Species) |>
  prep()
#> Error in `step_date()`:
#> Caused by error in `prep()`:
#> ✖ All columns selected for the step should be date or datetime.
#> • 2 double variables found: `Sepal.Length` and `Sepal.Width`
#> • 1 factor variable found: `Species`

recipe(~., data = ames) |>
  step_date(all_predictors()) |>
  prep()
#> Error in `step_date()`:
#> Caused by error in `prep()`:
#> ✖ All columns selected for the step should be date or datetime.
#> • 40 factor variables found: `MS_SubClass`, `MS_Zoning`, `Street`, …
#> • 12 double variables found: `Lot_Frontage`, `Mas_Vnr_Area`, …
#> • 22 integer variables found: `Lot_Area`, `Year_Built`, …
```

Comment (simonpcouch) [reply R/misc.R:466]: Heard! I'm with it, then.

File: tests/testthat/_snaps/ratio.md [hunk 6]
```diff
@@ -23,7 +25,8 @@
     Condition
       Error in `step_ratio()`:
       Caused by error in `prep()`:
-      ! All columns selected for the step should be double, or integer.
+      x All columns selected for the step should be double or integer.
+      * 8 factor variables found: `x5`, `x5`, `x5`, `x5`, ...
```
Comment (simonpcouch) [inline tests/testthat/_snaps/ratio.md:29]: What's happening here?

**Review by simonpcouch — APPROVED** [review]
Okay, looks good! Looks like there's one issue with duplicated column names; once that's resolved, thumbs up from me.

File: tests/testthat/_snaps/ratio.md (hunk 6 above)
Comment (EmilHvitfeldt) [reply tests/testthat/_snaps/ratio.md]: solved with a call to `unique()` [`0a36bb7` (#1235)](https://github.com/tidymodels/recipes/pull/1235/commits/0a36bb71440d49b5d96df6f481e6d367bee77418)
